README
App-Framework

A framework for making script development simpler....

(Proper documentation to follow)


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

EXAMPLES

The example scripts in examples/ show the expected use of App::Framework

	examples/eg1.pl	- Shows how to set up named arguments and the general structure of an application
	examples/eg2.pl	- Shows how to use options and how to drive the Sql API
	

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc App::Framework

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-Framework

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/App-Framework

    CPAN Ratings
        http://cpanratings.perl.org/d/App-Framework

    Search CPAN
        http://search.cpan.org/dist/App-Framework/


COPYRIGHT AND LICENCE

Copyright (C) 2009 Steve Price

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for App-Framework

1.06    Thu Mar 22 08:05:0 2012
		Fix: removed use of "defined(@array)" to eliminate Perl "deprecated" error message

1.05    Mon Aug 30 14:47:00 2011
		Minor fixes
		
1.04    Mon Nov 16 13:45:00 2009
		Bug fix: Corrected config path environment definition (see rt.cpan.org #49649)
		
1.03    Tue Sep 15 13:45:00 2009
		Bug fix: corrected Sql test (mysql user setting)
		Bug fix: Removed unnecessary debug output from Object
		
1.02    Tue Jul 28 15:45:00 2009
		Bug fix: Corrected pre-requisite PM list created in Makefile.PL
		Bug fix: corrected Run test (required programs checking)
		Added wildcard args support (for Windows)
		
1.01    Wed Jul 22 15:45:00 2009
		Bug fix: Corrected Run feature status checking
		Updated Mail feature & added documentation
		Modified Sql test to stop failures on some systems

1.00    Mon Jul 20 15:45:00 2009
		Release after user testing
		Bug fix: Corrected Run feature timeout capability
		Updated Run feature testing
		Re-worked options/args/data so that their "variables" can be used in any of the other sections
		Added logging feature

0.95    Wed Jun 10 14:04:00 2009
		Bug fix: Corrected various test scripts for other OSs
		Updated Makefile.PL to allow user to fully install optional modules with their dependencies
		Updated some more POD

0.94    Wed Jun 03 14:04:00 2009
		Fixed Filter extension
		Bug fix: Args defaults was not checking properly
		Updated Makefile.PL to check website for later version
		Updated some more POD

0.93    Fri May 22 14:04:00 2009
		Optimised debug & verbose methods in Object (was slugging apps)
		Corrected Extension & Filter modules
		Updated some more POD

0.92    Tue May 18 12:30:00 2009
		Tweak to Makefile to correct removal of old installation

0.91    Tue May 18 08:07:00 2009
		MANIFEST corrections

0.90    Mon May 18 12:30:00 2009
		Major re-write to make the framework modular. Now has core/extensions/features
		NOTE: Application 'run' subroutine renamed to 'app'

0.07    Sat Mar 14 20:33:00 2009
        Bug fix: Corrected Sql initialisation (was claiming DBI not installed)
        Added: -man-dev option to display app developer man pages
        Added: allow 'dev:' prefix to make option only show in developer man pages
        Added: Sql support for 'group by'

0.06    Thu Mar 05 15:09:00 2009
        Updated script pod. Modified so that scripts can be used without DBI. Corrected build fail when DBI not installed.

0.05    Wed Mar 04 13:29:00 2009
        Corrected MANIFEST problem that caused t/04-Args.t test to fail.

0.04    Wed Mar 04 08:07:00 2009
        Various bug fixes.

0.03    Tue Mar 03 15:55:56 2009
        Better testing, better documentation, various bug fixes.

0.02    Fri Feb 27 08:38:56 2009
        Fixed MANIFEST.SKIP (and MANIFEST).

0.01    Thu Feb 26 15:51:56 2009
        First version, testing distribution files are correct. This version has very limited testing.



examples/eg-sql.pl
#!/usr/bin/perl
#
use strict ;
use App::Framework '+Sql' ;

# VERSION
our $VERSION = '1.001' ;

	# Create application and run it
	App::Framework->new()->go() ;



#=================================================================================
# SUBROUTINES EXECUTED BY APP
#=================================================================================

#----------------------------------------------------------------------
# Main execution
#
sub app
{
	my ($app) = @_ ;
	
	# options
	my %opts = $app->options() ;

	# All named transactions use the values stored in this hash
	my %sql_vars ;

	# Set up database access
	$app->set(
		'sql' => {
			'database'	=> $opts{'database'},
			'table'		=> $opts{'table'},
			'user'		=> $opts{'user'},
			'password'	=> $opts{'password'},
			
			# Option to do some Sql debugging by tracing transactions to a file
			'trace'		=> $opts{'debug'} ? 4 : 0,
			'trace_file'=> 'logsql.log',
			'debug' => $opts{'debug'},
			
			# Prepare some named transactions
			'prepare'	=> {
				'check',  {
					'limit'	=> 1,
					'where'	=> {
						'vars'	=> [qw/pid channel/],
						'vals'	=> \%sql_vars,
					},
				},
				'select',  {
					'where'	=> {
						'sql' => '`pid`>=? AND `channel`=?',
						'vars'	=> [qw/pid channel/],
						'vals'	=> \%sql_vars,
					},
				},
				'select_group',  {
					'where'	=> {
						'sql' => '`pid`>=? AND `channel`=?',
						'vars'	=> [qw/pid channel/],
						'vals'	=> \%sql_vars,
					},
					'group'	=> 'channel',
				},
				'delete',  {
					'where'	=> {
						'sql' => '`pid`>=? AND `channel`=?',
						'vars'	=> [qw/pid channel/],
						'vals'	=> \%sql_vars,
					},
				},
				'insert',  {
					'vars'	=> [qw/pid channel title date start duration episode num_episodes repeat text/],
					'vals'	=> \%sql_vars,
				},
			},
		}
	) ;

	## demo of running sql stored as text in a __DATA__ section
	print "== Create table ===\n" ;
	$app->sql_from_data("listings2.sql") ;

	## Do some inserts
	my $pid = $$ ;
	my $start_pid = $pid ;
	my $start_chan = '_adummy' ;
	
	%sql_vars = (
		'pid'	=> $start_pid,
		'channel' => $start_chan,
		'title' => 'a test program',
		'date' => '2008-06-10',
		'start' => '10:30',
		'duration' => '01:00',
		'episode' => 2,
		'num_episodes' => 6,
		'repeat' => 0,
		'text' => "This is a test program",
	) ;

	print "Insert pid=$sql_vars{'pid'}...\n" ;
	$app->sql_query('insert') ;


	%sql_vars = (
		'pid'	=> ++$pid,
		'channel' => $start_chan,
		'title' => 'a test program 2',
		'date' => '2008-06-11',
		'start' => '11:30',
		'duration' => '01:00',
		'episode' => 2,
		'num_episodes' => 6,
		'repeat' => 0,
		'text' => "This is a test program",
	) ;
	print "Insert pid=$sql_vars{'pid'}...\n" ;
	$app->sql_query('insert') ;
	
	%sql_vars = (
		'pid'	=> ++$pid,
		'channel' => $start_chan,
		'title' => 'a test program 3',
		'date' => '2008-06-12',
		'start' => '12:30',
		'duration' => '01:00',
		'episode' => 2,
		'num_episodes' => 6,
		'repeat' => 0,
		'text' => "This is a test program",
	) ;
	print "Insert pid=$sql_vars{'pid'}...\n" ;
	$app->sql_query('insert') ;
	
	%sql_vars = (
		'pid'	=> ++$pid,
		'channel' => $start_chan,
		'title' => 'a test program 4',
		'date' => '2008-06-13',
		'start' => '13:30',
		'duration' => '01:00',
		'episode' => 2,
		'num_episodes' => 6,
		'repeat' => 0,
		'text' => "This is a test program",
	) ;
	print "Insert pid=$sql_vars{'pid'}...\n" ;
	$app->sql_query('insert') ;
	
	
	
	show($app, \%sql_vars, $start_pid, $start_chan, "Just stored...") ;
	show($app, \%sql_vars, 0, $start_chan, "All stored...") ;
	show($app, \%sql_vars, 0, $start_chan, "Grouped", 'select_group') ;
	
	print "Delete..\n" ;
	$sql_vars{'pid'} = $start_pid ;
	$app->sql_query('delete') ;
	
	show($app, \%sql_vars, 0, $start_chan, "After delete...") ;

}

#----------------------------------------------------------
sub show
{
	my ($app, $sql_vars_href, $pid, $chan, $title, $query_name) = @_ ;

	print "== $title =================\n" ;

	$query_name ||= 'select' ;
	
	## now get results back
	$sql_vars_href->{'pid'} = $pid ;
	$sql_vars_href->{'channel'} = $chan ;
	
	## demo a select transaction - could have done this as:
	# my @results = $app->sql_query_all('select');
	#
	$app->sql_query($query_name) ;
	while (my $href = $app->sql_next($query_name))
	{
		foreach my $key (sort keys %$href)
		{
			print "$key=$href->{$key} " ;
		}
		print "\n" ;
	}
	
}


#=================================================================================
# SETUP
#=================================================================================

__DATA__

[SUMMARY]
Tests the application object with SQL

[DESCRIPTION]

B<$name> will ensure that table 'listings2' is created in the specified database. It will then
add some rows, show them, and then delete them.


[OPTIONS]

-database=s	Database name [default=test]

Specify the database name to use

-table=s	Table name [default=listings2]

Specify a different table (note that this example will only ensure that table 'listings2' is created')

-user=s User

Your MySql user

-password=s	Pass

Your MySql password


__#=============================================================================================================================
__DATA__ listings2.sql

-- http://www.phpmyadmin.net
-- 
-- Host: localhost
-- Generation Time: Feb 26, 2009 at 12:13 PM
-- Server version: 5.0.51
-- PHP Version: 5.2.6
-- 
-- Database: `test`
-- 

-- --------------------------------------------------------

-- 
-- Table structure for table `listings2`
-- 

DROP TABLE IF EXISTS `listings2`;
CREATE TABLE IF NOT EXISTS `listings2` (
  `pid` varchar(128) NOT NULL,
  `title` varchar(128) NOT NULL,
  `date` date NOT NULL,
  `start` time NOT NULL,
  `duration` time NOT NULL,
  `episode` int(11) default NULL,
  `num_episodes` int(11) default NULL,
  `repeat` varchar(128) default NULL,
  `text` longtext NOT NULL,
  `channel` varchar(128) NOT NULL,
  `adapter` tinyint(8) NOT NULL default '0' COMMENT 'DVB adapter number',
  `genre` varchar(256) default NULL,
  `chan_type` varchar(256) NOT NULL default 'tv',
  `audio` tinyint(1) default '1',
  `video` tinyint(1) default '1',
  KEY `pid` (`pid`),
  KEY `chan_date_start_duration` (`channel`,`date`,`start`,`duration`)
) ENGINE=MyISAM DEFAULT CHARSET=latin1;

__#=============================================================================================================================
__END__